tests/integrationv2/common.py

```diff
@@ -111,7 +111,16 @@ def __init__(self, name, prefix, location=TEST_CERT_DIRECTORY):
             self.algorithm = 'RSAPSS'
 
     def compatible_with_cipher(self, cipher):
-        return (self.algorithm == cipher.algorithm) or (cipher.algorithm == 'ANY')
+        if self.algorithm == cipher.algorithm:
+            return True
+        if cipher.algorithm == 'ANY':
+            return True
+        if self.algorithm == 'RSAPSS':
+            if cipher.algorithm != 'RSA':
+                return False
+            if 'ECDHE' in cipher.name:
+                return True
```

Comment:
I assume this is the "hacky string comparisons", but I'm trying to understand the context here. Is RSA-PSS only compatible with ECDHE key exchange?

Reply:
RSA-PSS certs can't be used with RSA key exchange (that's what Sam was asking about [here](https://github.com/aws/s2n-tls/pull/4927#discussion_r1870294448)). That's because RSA-PSS certs can only be used for signing, not encryption, and RSA key exchange involves encryption. RSA-PSS certs are basically just RSA certs with safer usage rules.

But as I look at this again, checking for "ECDHE" is probably too restrictive. I'll relax it to "DHE". I can't go all the way to just checking for no "RSA", because legacy algorithms assume RSA without putting it in their names.

Reply:
Updated, and I updated the testing results in the description. And here's the result of only non-ECDHE:
```
$ cat signature_algorithms.out | grep ": PASSED " | grep pss_pss | grep TLS1.2 | grep -v ECDHE | wc -l
28
```

<details>

```
$ cat signature_algorithms.out | grep ": PASSED " | grep pss_pss | grep TLS1.2 | grep -v ECDHE
293: PASSED test_signature_algorithms.py::test_s2n_server_signature_algorithms[client-auth-rsa_pss_pss_sha256-RSA_PSS_2048_SHA256-TLS1.2-S2N-OpenSSL-DHE-RSA-AES128-SHA]
293: PASSED test_signature_algorithms.py::test_s2n_server_signature_algorithms[client-auth-rsa_pss_pss_sha256-RSA_PSS_2048_SHA256-TLS1.2-S2N-OpenSSL-DHE-RSA-AES256-SHA]
293: PASSED test_signature_algorithms.py::test_s2n_server_signature_algorithms[client-auth-rsa_pss_pss_sha256-RSA_PSS_2048_SHA256-TLS1.2-S2N-OpenSSL-DHE-RSA-AES128-GCM-SHA256]
293: PASSED test_signature_algorithms.py::test_s2n_server_signature_algorithms[client-auth-rsa_pss_pss_sha256-RSA_PSS_2048_SHA256-TLS1.2-S2N-OpenSSL-DHE-RSA-AES256-GCM-SHA384]
293: PASSED test_signature_algorithms.py::test_s2n_server_signature_algorithms[client-auth-rsa_pss_pss_sha256-RSA_PSS_2048_SHA256-TLS1.2-S2N-OpenSSL-DHE-RSA-AES128-SHA256]
293: PASSED test_signature_algorithms.py::test_s2n_server_signature_algorithms[client-auth-rsa_pss_pss_sha256-RSA_PSS_2048_SHA256-TLS1.2-S2N-OpenSSL-DHE-RSA-CHACHA20-POLY1305]
293: PASSED test_signature_algorithms.py::test_s2n_server_signature_algorithms[client-auth-rsa_pss_pss_sha256-RSA_PSS_2048_SHA256-TLS1.2-S2N-OpenSSL-DHE-RSA-AES256-SHA256]
293: PASSED test_signature_algorithms.py::test_s2n_server_signature_algorithms[no-client-auth-rsa_pss_pss_sha256-RSA_PSS_2048_SHA256-TLS1.2-S2N-OpenSSL-DHE-RSA-AES128-SHA]
293: PASSED test_signature_algorithms.py::test_s2n_server_signature_algorithms[no-client-auth-rsa_pss_pss_sha256-RSA_PSS_2048_SHA256-TLS1.2-S2N-OpenSSL-DHE-RSA-AES256-SHA]
293: PASSED test_signature_algorithms.py::test_s2n_server_signature_algorithms[no-client-auth-rsa_pss_pss_sha256-RSA_PSS_2048_SHA256-TLS1.2-S2N-OpenSSL-DHE-RSA-AES128-SHA256]
293: PASSED test_signature_algorithms.py::test_s2n_server_signature_algorithms[no-client-auth-rsa_pss_pss_sha256-RSA_PSS_2048_SHA256-TLS1.2-S2N-OpenSSL-DHE-RSA-AES256-SHA256]
293: PASSED test_signature_algorithms.py::test_s2n_server_signature_algorithms[no-client-auth-rsa_pss_pss_sha256-RSA_PSS_2048_SHA256-TLS1.2-S2N-OpenSSL-DHE-RSA-AES128-GCM-SHA256]
293: PASSED test_signature_algorithms.py::test_s2n_server_signature_algorithms[no-client-auth-rsa_pss_pss_sha256-RSA_PSS_2048_SHA256-TLS1.2-S2N-OpenSSL-DHE-RSA-AES256-GCM-SHA384]
293: PASSED test_signature_algorithms.py::test_s2n_server_signature_algorithms[no-client-auth-rsa_pss_pss_sha256-RSA_PSS_2048_SHA256-TLS1.2-S2N-OpenSSL-DHE-RSA-CHACHA20-POLY1305]
293: PASSED test_signature_algorithms.py::test_s2n_client_signature_algorithms[client-auth-rsa_pss_pss_sha256-RSA_PSS_2048_SHA256-TLS1.2-S2N-OpenSSL-DHE-RSA-AES128-SHA]
293: PASSED test_signature_algorithms.py::test_s2n_client_signature_algorithms[client-auth-rsa_pss_pss_sha256-RSA_PSS_2048_SHA256-TLS1.2-S2N-OpenSSL-DHE-RSA-AES256-SHA]
293: PASSED test_signature_algorithms.py::test_s2n_client_signature_algorithms[client-auth-rsa_pss_pss_sha256-RSA_PSS_2048_SHA256-TLS1.2-S2N-OpenSSL-DHE-RSA-AES128-SHA256]
293: PASSED test_signature_algorithms.py::test_s2n_client_signature_algorithms[client-auth-rsa_pss_pss_sha256-RSA_PSS_2048_SHA256-TLS1.2-S2N-OpenSSL-DHE-RSA-AES256-SHA256]
293: PASSED test_signature_algorithms.py::test_s2n_client_signature_algorithms[client-auth-rsa_pss_pss_sha256-RSA_PSS_2048_SHA256-TLS1.2-S2N-OpenSSL-DHE-RSA-AES128-GCM-SHA256]
293: PASSED test_signature_algorithms.py::test_s2n_client_signature_algorithms[client-auth-rsa_pss_pss_sha256-RSA_PSS_2048_SHA256-TLS1.2-S2N-OpenSSL-DHE-RSA-AES256-GCM-SHA384]
293: PASSED test_signature_algorithms.py::test_s2n_client_signature_algorithms[client-auth-rsa_pss_pss_sha256-RSA_PSS_2048_SHA256-TLS1.2-S2N-OpenSSL-DHE-RSA-CHACHA20-POLY1305]
293: PASSED test_signature_algorithms.py::test_s2n_client_signature_algorithms[no-client-auth-rsa_pss_pss_sha256-RSA_PSS_2048_SHA256-TLS1.2-S2N-OpenSSL-DHE-RSA-AES128-SHA]
293: PASSED test_signature_algorithms.py::test_s2n_client_signature_algorithms[no-client-auth-rsa_pss_pss_sha256-RSA_PSS_2048_SHA256-TLS1.2-S2N-OpenSSL-DHE-RSA-AES256-SHA]
293: PASSED test_signature_algorithms.py::test_s2n_client_signature_algorithms[no-client-auth-rsa_pss_pss_sha256-RSA_PSS_2048_SHA256-TLS1.2-S2N-OpenSSL-DHE-RSA-AES128-SHA256]
293: PASSED test_signature_algorithms.py::test_s2n_client_signature_algorithms[no-client-auth-rsa_pss_pss_sha256-RSA_PSS_2048_SHA256-TLS1.2-S2N-OpenSSL-DHE-RSA-AES256-SHA256]
293: PASSED test_signature_algorithms.py::test_s2n_client_signature_algorithms[no-client-auth-rsa_pss_pss_sha256-RSA_PSS_2048_SHA256-TLS1.2-S2N-OpenSSL-DHE-RSA-AES128-GCM-SHA256]
293: PASSED test_signature_algorithms.py::test_s2n_client_signature_algorithms[no-client-auth-rsa_pss_pss_sha256-RSA_PSS_2048_SHA256-TLS1.2-S2N-OpenSSL-DHE-RSA-AES256-GCM-SHA384]
293: PASSED test_signature_algorithms.py::test_s2n_client_signature_algorithms[no-client-auth-rsa_pss_pss_sha256-RSA_PSS_2048_SHA256-TLS1.2-S2N-OpenSSL-DHE-RSA-CHACHA20-POLY1305]
```

</details>

Reply:
Very interesting. Thank you for explaining!